streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java

```diff
@@ -283,7 +293,12 @@ public void init(final StateStoreContext context, final StateStore root) {
 
     // VisibleForTesting
     void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> records) {
-        // advance stream time to the max timestamp in the batch
+        // copy the observed stream time, for use in deciding whether to drop records during restore,
+        // when records have exceeded the store's grace period.
+        long streamTimeForRestore = observedStreamTime;
```

Comment:
Wondering if this would be correct?

If we have `st = 100`, `grace=10` and we do `put(k,v,95)` the put is correct. If we restore at `st=110`, the would still need to keep `k,v` and not drop it, even if it's timestamp 95 is now "too old"?

Reply:
Yeah this logic is pretty nuanced. (I tried to clarify in the comments but evidently not successfully.)

The `doPut()` method is not responsible for deciding when a put is too old (according to grace period); that check happens before `doPut()` is called. Inside the `doPut()` method, however, `observedStreamTime` is still used to decide when old records have fallen out of history retention. If a record has fallen out of history retention, then we don't need to keep it in the store, and therefore `doPut()` returns.

In this restore logic here, `streamTimeForRestore` is used to perform the grace period check. It would be incorrect to advance `streamTimeForRestore` at once for the entire batch, for the reason you gave above. In your example, we do still want to call `doPut()` for the record with `ts=95`. Assuming that is the first record in the restore batch, then `streamTimeForRestore=100` so `ts=95` and we call `doPut()` as we should. Only once we reach the later records in the restore batch will `streamTimeForRestore` be advanced past 100. 

OTOH, `observedStreamTime` can be advanced to the end of the batch right away. This allows us to optimize situations where, for example, a record near the beginning of the restore batch which we would put into the store would be immediately expired (based on history retention) by the end of the restore batch, and therefore we can skip putting it in inside `doPut()`. Here's an example:
* stream time is 50 at the start of the restore batch
* segment interval is 25
* stream time will be 100 by the end of the restore batch
* restore batch contains a record `(k, v, 50)` and also `(k, v, 60)`.

During restore when we see `(k, v, 50)`, we have to put it into the store (it's the latest value for the key so far). Then when we see `(k, v, 60)`, we also have to put it into the store (it's the new latest value) but we do NOT have to move `(k, v, 50)` into a segment store, because the segment that it would be moved into will be expired by the end of the restore process.

Here's another example: exact same as above, but the restore batch contains `(k, v, 60)` before `(k, v, 50)`, instead of after. When we see `(k, v, 60)` we have to put it into the store. When we see `(k, v, 50)`, we still call `doPut()` because it's not expired based on grace period, but `doPut()` will see that it is expired based on history retention (using `observedStreamTime=100`, the value it will be by the end of the restore batch) and therefore `doPut()`returns without inserting into the store.

Reply:
I guess the question is, what is the value of `observedStreamTime` when we start the restore? Are you saying it's `-1` and we basically "reply" `observedStreamTime` during restore? I guess I got confused with "streamTime" that is tracked by KS runtime and preserved across restarts; but the store does not use it (IIRC), but rather tracks its own time, right?

Maybe best to update some variable names? In the end, we do a "real reply" of stream-time for "grace period", and we apply an optimization for "history retention" by looking ahead (to the end of the batch) -> `endOfBatchStreamTime`. -- I guess follow up work (independent for this KIP) might be, to actually make use of KS runtime streamTime instead of tracking inside the store, and thus won't need `observedStreamTime` any longer, as we could look ahead to the "end-of-restore stream-time" (not just "end-of batch").

Reply:
> I guess the question is, what is the value of `observedStreamTime` when we start the restore? Are you saying it's `-1` and we basically "replay" `observedStreamTime` during restore?

Yes, that's exactly right. `observedStreamTime` is tracked locally per store. It is initialized to `-1` and only updated on `put()` or during restore. (This is the same as the existing behavior for window stores today.)

> Maybe best to update some variable names?

Are you proposing that `doPut()` takes stream time as a parameter, so that during normal `put()` operation we pass `observedStreamTime` and during restore we pass `endOfBatchStreamTime`, which means we can rename `streamTimeForRestore` to be `observedStreamTime` instead? This SGTM, just want to check whether that's also what you have in mind, since we removed a number of parameters from `doPut()` in a previous PR revision in order to keep the parameter list small.

> I guess follow up work (independent for this KIP) might be, to actually make use of KS runtime streamTime instead of tracking inside the store, and thus won't need `observedStreamTime` any longer, as we could look ahead to the "end-of-restore stream-time" (not just "end-of batch").

What's the scope of the "streamTime" which is tracked by the KS runtime? Is it per-task? Per-processor? Global? I'm wondering how this would work in situations with multiple partitions, or with multiple processors where some processors are expected to see new data earlier than other (downstream) processors.

I guess we'd also need to implement the change from your other comment about not writing records which are expired (based on grace period) into the changelog topic first before we can make this change, otherwise we would not have a way to determine during restore whether records are expired or not.

Reply:
Did not have a concrete proposal. Should be fine I guess.

Currently, `streamTime` is tracked per task (based on input records over all partitions). And yes, there is all kind of tricky things that you call out. Even if we have a filter() downstream processors see only a subset of data and their "internal stream-time (if they have any)" could be different (ie lagging). Caching has a similar effect.

There is a proposal to let KS track streamTime per processor, too.

Bottom line: it's complicated and need proper design and a KIP by itself...